cmd/hook.go

```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
+	}
+
+	const VersionHead string = "version=1"
+
+	if !strings.HasPrefix(rs.Data, VersionHead) {
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
+	hasPushOptions := false
+	response := []byte(VersionHead)
+	if strings.Contains(rs.Data, "push-options") {
+		response = append(response, byte(0))
+		response = append(response, []byte("push-options")...)
+		hasPushOptions = true
+	}
+	response = append(response, []byte("\n")...)
+
+	rs, err = readPktLine(reader)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+	}
+	if rs.Type != pktLineTypeFlush {
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
+	err = writePktLine(os.Stdout, pktLineTypeData, response)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+	}
+
+	err = writePktLine(os.Stdout, pktLineTypeFlush, nil)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+	}
+
+	// 2. receive commands from server.
+	// S: PKT-LINE(<old-oid> <new-oid> <ref>)
+	// S: ... ...
+	// S: flush-pkt
+	// # receive push-options
+	// S: PKT-LINE(push-option)
+	// S: ... ...
+	// S: flush-pkt
+	hookOptions := private.HookOptions{
+		UserName: pusherName,
+		UserID:   pusherID,
+	}
+	hookOptions.OldCommitIDs = make([]string, 0, hookBatchSize)
+	hookOptions.NewCommitIDs = make([]string, 0, hookBatchSize)
+	hookOptions.RefFullNames = make([]string, 0, hookBatchSize)
+
+	for {
+		rs, err = readPktLine(reader)
+		if err != nil {
+			fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+		}
+		if rs.Type == pktLineTypeFlush {
+			break
+		}
+		t := strings.SplitN(rs.Data, " ", 3)
+		if len(t) != 3 {
+			continue
+		}
+		hookOptions.OldCommitIDs = append(hookOptions.OldCommitIDs, t[0])
+		hookOptions.NewCommitIDs = append(hookOptions.NewCommitIDs, t[1])
+		hookOptions.RefFullNames = append(hookOptions.RefFullNames, t[2])
+	}
+
+	hookOptions.GitPushOptions = make(map[string]string)
+
+	if hasPushOptions {
+		for {
+			rs, err = readPktLine(reader)
+			if err != nil {
+				fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+			}
+			if rs.Type == pktLineTypeFlush {
+				break
+			}
+
+			kv := strings.SplitN(rs.Data, "=", 2)
+			if len(kv) == 2 {
+				hookOptions.GitPushOptions[kv[0]] = kv[1]
+			}
+		}
+	}
+
+	// run hook
+	resp, err := private.HookProcReceive(repoUser, repoName, hookOptions)
+	if err != nil {
+		fail("Internal Server Error", "run proc-receive hook failed :%v", err)
+	}
+
+	// 3 response result to service.
+	// # OK, but has an alternate reference.  The alternate reference name
+	// # and other status can be given in option directives.
+	// H: PKT-LINE(ok <ref>)
+	// H: PKT-LINE(option refname <refname>)
+	// H: PKT-LINE(option old-oid <old-oid>)
+	// H: PKT-LINE(option new-oid <new-oid>)
+	// H: PKT-LINE(option forced-update)
+	// H: ... ...
+	// H: flush-pkt
+	for _, rs := range resp.Results {
```

Comment:
There are 4 types of responses supported from hook to receive-pack, but you only implement one of them.

1. OK, run this command successfully.

        H: PKT-LINE(ok <ref>)

2. NO, I reject it.

        H: PKT-LINE(ng <ref> <reason>)

3. Fall through, let 'receive-pack' to execute it.

        H: PKT-LINE(ok <ref>)
        H: PKT-LINE(option fall-through)

4. OK, but has an alternate reference.  The alternate reference name and other status can be given in option directives.

        H: PKT-LINE(ok <ref>)
        H: PKT-LINE(option refname <refname>)
        H: PKT-LINE(option old-oid <old-oid>)
        H: PKT-LINE(option new-oid <new-oid>)
        H: PKT-LINE(option forced-update)

Reply:
I know it, but other types is not usefull now.
@jiangxin By the way, I have a small question, has any way to check if user push with ``-f`` option in this hook? Thanks.

Reply:
I think we need to ensure that:

```
H: PKT-LINE(ok <ref>) 
H: PKT-LINE(option fall-through)
```

is sent if we're sent a reference that isn't matching the ones we expect.